### PR TITLE
Add ALPN protocol selection

### PR DESCRIPTION
**Pull request by javalikescript**
ALPN protocol selection is required for HTTP/2

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -981,6 +981,140 @@ static int openssl_ssl_ctx_set_cert_verify(lua_State*L)
   return 0;
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10002000L
+/***
+set the list of ALPN protocols available to be negotiated
+@function set_alpn_protos
+@tparam[opt] table protos the protocol list
```
Comment (zhaozg) [inline src/ssl.c:988]: @tparam table protos the protocol list
@tparam function callback

This api can accept two paramaters, arg 1 is protos table, arg 2 is callback function. One time lua call wrap two OpenSSL APIs, Lua should make that simple.

**Comment by zhaozg** [comment]
And please add some `lua test code` for this feature, if you can.

File: src/ssl.c [hunk 2]
```diff
@@ -1676,6 +1810,10 @@ static luaL_Reg ssl_ctx_funcs[] =
 #endif
   {"verify_mode",     openssl_ssl_ctx_verify_mode},
   {"set_cert_verify", openssl_ssl_ctx_set_cert_verify},
+#if OPENSSL_VERSION_NUMBER >= 0x10002000L
+  {"set_alpn_protos",    openssl_ssl_ctx_set_alpn_protos},
+  {"set_alpn_select_cb", openssl_ssl_ctx_set_alpn_select_cb},
```
Comment (zhaozg) [inline src/ssl.c:1815]: `set_alpn_protos` and `set_alpn_select_cb` can be merge into one.

File: src/ssl.c [hunk 3]
```diff
@@ -2726,6 +2864,26 @@ static int openssl_ssl_tostring(lua_State*L)
   return 1;
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10002000L
+/***
+get the ALPN protocol selected
+@treturn the ALPN protocol selected or nil
+@function get0_alpn_selected
+*/
+static int openssl_ssl_get0_alpn_selected(lua_State*L)
```
[the remark this answers is not in the document]
Comment (javalikescript) [reply src/ssl.c]: Method renamed in commit 425bb693fc5fe814784761b6155653c713a7ca2c

File: src/ssl.c [hunk 4]
```diff
@@ -2769,6 +2927,9 @@ static luaL_Reg ssl_funcs[] =
   {"renegotiate_pending",   openssl_ssl_renegotiate_pending},
   {"set_connect_state",     openssl_ssl_set_connect_state},
   {"set_accept_state",      openssl_ssl_set_accept_state},
+#if OPENSSL_VERSION_NUMBER >= 0x10002000L
+  {"get0_alpn_selected",    openssl_ssl_get0_alpn_selected },
+#endif
```
[the remark this answers is not in the document]
Comment (javalikescript) [reply src/ssl.c:2932]: Method renamed in commit 425bb693fc5fe814784761b6155653c713a7ca2c

File: src/ssl.c (hunk 1 above)
Comment (javalikescript) [reply src/ssl.c]: As far as I understood, the protocol list is provided by the client and the selection callback is used on the server side.
I would recommend to stick to the openssl two methods and naming but do not hesitate to indicate if you still want one method.

Comment (zhaozg) [reply src/ssl.c]: It's ok

**Comment by zhaozg** [comment]
Thanks your work.

**Pull request merged**